src/gpu/tsvd/tsvd.cu

```diff
@@ -227,6 +229,195 @@ void calculate_u(const Matrix<float> &X, const Matrix<float> &Q, const Matrix<fl
+ *
+ * @param X
+ * @param _Q
+ * @param _w
+ * @param _U
+ * @param _explained_variance
+ * @param _explained_variance_ratio
+ * @param _param
+ */
+void cusolver_tsvd(Matrix<float> &X, double* _Q, double* _w, double* _U, double* _explained_variance, double* _explained_variance_ratio, params _param){
+	//Allocate matrix for X^TX
+	Matrix<float>XtX(_param.X_n, _param.X_n);
+
+	//Create context
+	DeviceContext context;
+
+	//Multiply X and Xt and output result to XtX
+	multiply(X, X, XtX, context, true, false, 1.0f);
+
+	//Set up Q (V^T) and w (singular value) matrices (w is a matrix of size Q.rows() by 1; really just a vector
+	Matrix<float>Q(XtX.rows(), XtX.columns()); // n X n -> V^T
+	Matrix<float>w(Q.rows(), 1);
+	calculate_eigen_pairs_exact(XtX, Q, w, context);
+
+	//Get tsvd attributes
+	get_tsvd_attr(X, Q, _Q, w, _w, _U, _explained_variance, _explained_variance_ratio, _param, context);
+}
+
+/**
+ * Conduct truncated svd using the power method
+ *
+ * @param X
+ * @param _Q
+ * @param _w
+ * @param _U
+ * @param _explained_variance
+ * @param _explained_variance_ratio
+ * @param _param
+ */
+void power_tsvd(Matrix<float> &X, double* _Q, double* _w, double* _U, double* _explained_variance, double* _explained_variance_ratio, params _param){
+	//Allocate matrix for X^TX
+	Matrix<float>M(_param.X_n, _param.X_n);
+
+	//Allocate matrix to be used in cublasSger outer product calculation
+	Matrix<float>A(M.rows(), M.columns());
+
+	//Create context
+	DeviceContext context;
+
+	//Multiply X and Xt and output result to XtX
+	multiply(X, X, M, context, true, false, 1.0f);
+
+//	//Temporary matrix whcih will be updated in power method iteratively
+//    Matrix<float>M(XtX.rows(), XtX.columns());
+//	M.copy(XtX);
+
+	//Set up Q (V^T) and w (singular value) matrices (w is a matrix of size Q.rows() by 1; really just a vector
+	Matrix<float>Q(M.rows(), _param.k);
+	Matrix<float>w(_param.k, 1);
+	std::vector<double> w_temp(w.size());
+
+	/* Power Method for finding all eigenvectors/eigen values
+	 * Logic:
+	 *
+	 * Since the matrix is symmetric, there exists an orthogonal basis of eigenvectors. Once you have found an eigenvector, extend it to an orthogonal basis,
+	 * rewrite the matrix in terms of this basis and restrict to the orthogonal space of the known eigenvector.
+	 * This is comprised in the method of deflation:
+	 *     You first find the first eigenvector v1 (with a maximal λ1), by iterating xn+1=Axn/|Axn|with a "random" initial x0. Once you have found a good approximation
+	 *     for v1, you consider B=A−λ1|v1|2 * v1vT (this simple step replaces the "rewrite in terms of this basis" above). This is a matrix that behaves like A for anything
+	 *     orthogonal to v1 and zeroes out v1. Use the power method on B again, which will reveal v2, an eigenvector of a largest eigenvalue of B.
+	 *     Then switch to C=B−λ2|v2|2 * v2vT so on.
+	 */
+	Matrix<float>b_k(_param.X_n, 1);
+	Matrix<float>b_k1(_param.X_n, 1);
+	float tol = _param.tol;
```

Comment:
tol is an unnecessary variable. Just directly use _param.tol.

Reply:
Yep, not needed.